sqlflow_models/deep_embedding_cluster.py

```diff
@@ -0,0 +1,295 @@
+        def _concate_generate(dataset_element, label):
+            concate_y = tf.stack([dataset_element[feature.key] for feature in self._feature_columns], axis=1)
+            return (dataset_element, concate_y)
+
+        y = x.map(map_func=_concate_generate)
+        y.prefetch(1)
+        print('{} Finished dataset transform.'.format(datetime.now()))
+
+        # Layers - decoder
+        self.decoder_layers = []
+        for i in range(self._n_stacks - 1, 0, -1):
+            self.decoder_layers.append(Dense(units=self._pretrain_dims[i],
+                                             activation=self._pretrain_activation_func,
+                                             kernel_initializer=self._pretrain_initializer,
+                                             name='decoder_%d' % i))
+
+        self.decoder_layers.append(Dense(units=self._pretrain_dims[0],
+                                         kernel_initializer=self._pretrain_initializer,
+                                         name='decoder_0'))
+        # Pretrain - autoencoder, encoder
+        # autoencoder
+        self._autoencoder = keras.Sequential(layers=[self.input_layer] + self.encoder_layers + self.decoder_layers,
+                                             name='autoencoder')
+        self._autoencoder.compile(optimizer=self._pretrain_optimizer, loss='mse')
+        # encoder
+        self._encoder = keras.Sequential(layers=[self.input_layer] + self.encoder_layers, name='encoder')
+        self._encoder.compile(optimizer=self._pretrain_optimizer, loss='mse')
+
+        callbacks = [
+            EarlyStopping(monitor='loss', patience=2, min_delta=0.001),
+            ReduceLROnPlateau(monitor='loss', factor=0.1, patience=2)
+        ]
+        print('{} Training auto-encoder.'.format(datetime.now()))
+        self._autoencoder.fit_generator(generator=y, epochs=self._pretrain_epochs, callbacks=callbacks)
+
+        # encoded_input
+        # type : numpy.ndarray shape : (num_of_all_records,num_of_cluster) (70000,10) if mnist
+        print('{} Calculating encoded_input.'.format(datetime.now()))
+        self.encoded_input = self._encoder.predict(x)
+
+        del self._autoencoder
+        del self._encoder
+        del self.decoder_layers
+        print('{} Done pre-train.'.format(datetime.now()))
+
+    def call(self, inputs, training=None, mask=None):
+        x = self.input_layer(inputs)
+        for encoder_layer in self.encoder_layers:
+            x = encoder_layer(x)
+        return self.clustering_layer(x)
+
+    def init_centroids(self):
+        """
+        Training K-means `_kmeans_init` times on the output of encoder to get best initial centroids.
+        :return:
+        """
+        self.kmeans = KMeans(n_clusters=self._n_clusters, n_init=self._kmeans_init)
+        self.y_pred_last = self.kmeans.fit_predict(self.encoded_input)
+        print('{} Done init centroids by k-means.'.format(datetime.now()))
+
+    def cluster_train_loop(self, x):
```

Comment:
Maybe we can make `cluster_train_loop` to be more general: 
``` sql
def sqlflow_trai_loop(self, dataset, epochs, verbos):
    ...
```

Reply:
> Maybe we can make `cluster_train_loop` to be more general:
> 
> ```sql
> def sqlflow_trai_loop(self, dataset, epochs, verbos):
>     ...
> ```

Actually I have thought about this.  
`epochs` is always used to define how many times to train on the whole dataset. However in this case, the auxiliary target distribution `p` needs to be updated after `update_interval` iterations  (`update_interval` always smaller than the number of batches in one `epoch`). Since the pre-train process can fetch parameter `pretrain_epochs` from construct function, and `pre_train` should be consider as sub-train-process, I did not add `epochs` as parameter here.
`verbose` can be add here for control to display more details. I can add this later.

Reply:
Thanks, I see that `epochs` is not suitable for this cause, but we need to support the custom train loop function  by editing `template_tf.go` in SQLFlow, and I think `epochs` is useful for the most of the scenes, so maybe we can to add `epochs` argument with a default value.